pkg/costattribution/tracker.go

```diff
@@ -0,0 +1,363 @@
+	slices.Sort(orderedLables)
+
+	// Create a map for overflow labels to export when overflow happens
+	overflowLabels := make([]string, len(orderedLables)+2)
+	for i := range orderedLables {
+		overflowLabels[i] = overflowValue
+	}
+
+	overflowLabels[len(orderedLables)] = userID
+	overflowLabels[len(orderedLables)+1] = overflowValue
+
+	tracker := &Tracker{
+		userID:                  userID,
+		labels:                  orderedLables,
+		maxCardinality:          limit,
+		observed:                make(map[string]*observation),
+		hashBuffer:              make([]byte, 0, 1024),
+		cooldownDuration:        cooldown,
+		logger:                  logger,
+		overflowLabels:          overflowLabels,
+		totalFailedActiveSeries: atomic.NewFloat64(0),
+	}
+
+	variableLabels := slices.Clone(orderedLables)
+	variableLabels = append(variableLabels, tenantLabel, "reason")
+	tracker.discardedSampleAttribution = prometheus.NewDesc("cortex_discarded_attributed_samples_total",
+		"The total number of samples that were discarded per attribution.",
+		variableLabels,
+		prometheus.Labels{trackerLabel: defaultTrackerName})
+
+	tracker.receivedSamplesAttribution = prometheus.NewDesc("cortex_received_attributed_samples_total",
+		"The total number of samples that were received per attribution.",
+		variableLabels[:len(variableLabels)-1],
+		prometheus.Labels{trackerLabel: defaultTrackerName})
+
+	tracker.activeSeriesPerUserAttribution = prometheus.NewDesc("cortex_ingester_attributed_active_series",
+		"The total number of active series per user and attribution.", variableLabels[:len(variableLabels)-1],
+		prometheus.Labels{trackerLabel: defaultTrackerName})
+
+	return tracker
+}
+
+func (t *Tracker) hasSameLabels(labels []string) bool {
+	return slices.Equal(t.labels, labels)
+}
+
+var bufferPool = sync.Pool{
+	New: func() interface{} {
+		return new(bytes.Buffer)
+	},
+}
+
+func (t *Tracker) cleanupTrackerAttribution(key string) {
+	t.observedMtx.Lock()
+	defer t.observedMtx.Unlock()
+	delete(t.observed, key)
+}
+
+func (t *Tracker) IncrementActiveSeries(lbs labels.Labels, now time.Time) {
+	if t == nil {
+		return
+	}
+	t.updateCounters(lbs, now, 1, 0, 0, nil, true)
+}
+
+func (t *Tracker) DecrementActiveSeries(lbs labels.Labels) {
+	if t == nil {
+		return
+	}
+	t.updateCounters(lbs, time.Time{}, -1, 0, 0, nil, false)
+}
+
+func (t *Tracker) Collect(out chan<- prometheus.Metric) {
+
+	if t.isOverflow.Load() {
+		out <- prometheus.MustNewConstMetric(t.activeSeriesPerUserAttribution, prometheus.GaugeValue, t.overflowCounter.activeSerie.Load(), t.overflowLabels[:len(t.overflowLabels)-1]...)
+		out <- prometheus.MustNewConstMetric(t.receivedSamplesAttribution, prometheus.CounterValue, t.overflowCounter.receivedSample.Load(), t.overflowLabels[:len(t.overflowLabels)-1]...)
+		out <- prometheus.MustNewConstMetric(t.discardedSampleAttribution, prometheus.CounterValue, t.overflowCounter.totalDiscarded.Load(), t.overflowLabels...)
```

Comment:
Why are we doing the sub-slicing here to the length of the slice? That sounds like a noop.

Reply:
Since the discarded sample metric includes an additional "reason" label compared to the other two, it is the only one that contains the full set of overflow labels. The other two metrics will have one fewer label (i.e., len(overflow)-1).